java-dynamic-sqs-listener-api/src/main/java/com/jashmore/sqs/argument/ArgumentResolver.java

```diff
@@ -0,0 +1,41 @@
+package com.jashmore.sqs.argument;
+
+import com.amazonaws.services.sqs.model.Message;
+import com.jashmore.sqs.QueueProperties;
+import com.jashmore.sqs.broker.MessageBroker;
+import com.jashmore.sqs.processor.MessageProcessor;
+
+import java.lang.reflect.Parameter;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Used to determine if it can fulfill resolving a method parameter from the content of a message.
+ *
+ * <p>If you were to consider this library as similar to a pub-sub system, this could be considered the bus.  It polls for messages from the
+ * remote queue and will pass them to the {@link MessageBroker} which will delegate it to the {@link MessageProcessor} that knows how to process
+ * this message.
+ *
+ * <p>As there could be multiple messages all being processed at once, and therefore resolving many arguments concurrently,
+ * the implementations of this class must be thread safe.
+ */
+@ThreadSafe
```

Comment:
💯 on `ThreadSafe` and the accompanying docs around that and what the implementers are expected to do.